com.unity.ml-agents.extensions/Runtime/Match3/Match3Actuator.cs

```diff
@@ -37,65 +34,73 @@ public class Match3Actuator : IActuator, IHeuristicProvider, IBuiltInActuator
+            var numMoves = Move.NumPotentialMoves(m_MaxBoardSize);
             m_ActionSpec = ActionSpec.MakeDiscrete(numMoves);
             m_Random = new System.Random(seed);
         }
 
         /// <inheritdoc/>
         public ActionSpec ActionSpec => m_ActionSpec;
 
+        [Conditional("DEBUG")]
+        void CheckBoardSizes()
+        {
+            var currentBoardSize = m_Board.GetCurrentBoardSize();
+            if (currentBoardSize >= m_MaxBoardSize)
+            {
+                Debug.LogWarning(
+                    "Current BoardSize is larger than maximum board size was on initialization. This may cause unexpected results.\n" +
+                    $"Original GetMaxBoardSize() result: {m_MaxBoardSize}\n" +
+                    $"GetCurrentBoardSize() result: {currentBoardSize}"
+                );
+            }
+        }
+
         /// <inheritdoc/>
         public void OnActionReceived(ActionBuffers actions)
         {
+            CheckBoardSizes();
             if (m_ForceHeuristic)
             {
                 Heuristic(actions);
             }
             var moveIndex = actions.DiscreteActions[0];
 
-            if (m_Board.Rows != m_Rows || m_Board.Columns != m_Columns || m_Board.NumCellTypes != m_NumCellTypes)
-            {
-                Debug.LogWarning(
-                    $"Board shape changes since actuator initialization. This may cause unexpected results. " +
-                    $"Old shape: Rows={m_Rows} Columns={m_Columns}, NumCellTypes={m_NumCellTypes} " +
-                    $"Current shape: Rows={m_Board.Rows} Columns={m_Board.Columns}, NumCellTypes={m_Board.NumCellTypes}"
-                );
-            }
-
-            Move move = Move.FromMoveIndex(moveIndex, m_Rows, m_Columns);
+            Move move = Move.FromMoveIndex(moveIndex, m_MaxBoardSize);
             m_Board.MakeMove(move);
         }
 
         /// <inheritdoc/>
         public void WriteDiscreteActionMask(IDiscreteActionMask actionMask)
         {
+            var currentBoardSize = m_Board.GetCurrentBoardSize();
+            CheckBoardSizes();
             const int branch = 0;
             bool foundValidMove = false;
             using (TimerStack.Instance.Scoped("WriteDiscreteActionMask"))
             {
                 var numMoves = m_Board.NumMoves();
 
-                var currentMove = Move.FromMoveIndex(0, m_Board.Rows, m_Board.Columns);
+                var currentMove = Move.FromMoveIndex(0, m_MaxBoardSize);
                 for (var i = 0; i < numMoves; i++)
                 {
-                    if (m_Board.IsMoveValid(currentMove))
+                    // Check that the move is allowed for the current boardSize (e.g. it won't move a piece out of
+                    // bounds), and that it's allowed by the game itself.
+                    if (currentMove.IsValidForBoardSize(currentBoardSize) && m_Board.IsMoveValid(currentMove))
```

Comment:
this reads a bit clunky, couldn't the board check if the move would do something invalid based on its current size?